### PR TITLE
types.jl: sleep with Date.Period types

### LINKED ISSUE #19736: `sleep`, etc. with `Dates.Period` types
It would be nice for standard library unity if this were supported:

```julia
sleep(Dates.Second(10))
sleep(Dates.Millisecond(10))
```

I personally have a hard time remembering whether sleep takes units of seconds or milliseconds.

The same goes for functions like `timedwait`.

### PR DESCRIPTION
**Pull request by Shade5**
Supports sleep(Dates.Second(10))
		 sleep(Dates.Milliesecond(10))

Closes #19736

### CODE REVIEW

**Review by TotalVerb — CHANGES_REQUESTED** [review]
Thanks for the contribution! A few comments:

As is, this will only work for seconds and milliseconds. That's what I asked for, but it can be made more generic. There's a `Dates.toms` function that takes any fixed-length `Dates.Period` and gives the number of milliseconds it represents. I would therefore suggest writing

```julia
sleep(time::Dates.Period) = sleep(Dates.toms(time) / 1000)
```

Also, in addition to `sleep`, there are functions like `Timer`, `timedwait` that could receive the same treatment.

I also think that this is not the right file to define these methods in. Could these methods be moved near the other methods for these functions?

As for tests, `sleep` is not easy to test properly. But I think having a test that the call does not raise an error is useful. One could do

```julia
@test sleep(Dates.Millisecond(10)) === nothing
```

Finally, make sure the file ends in a new line character. If it doesn't, it's likely that Travis will reject the PR.

**Comment by Shade5** [comment]
@TotalVerb I'll make it more generic as you suggested and also include timer and timedwait.
I tried defining these new functions near the old ones but the problem is that while building julia the Date module is not defined yet while the time modules are being built. Thus I had to move it to the Date module

**Comment by TotalVerb** [comment]
You could consider moving the loading of `Dates` in `sysimg.jl`, line 357, a little earlier. Is there a reason why `Dates` is loaded so late? cc @quinnj

**Comment by TotalVerb** [comment]
Note that the `repeat` argument for `Timer` is also a time.

**Comment by Shade5** [comment]
So, I'll have to add three more functions
```
Timer(time::Dates.Period, repeat::Real=0.0)
Timer(time::Real, repeat::Dates.Period=0.0)
Timer(time::Dates.Period, repeat::Dates.Period=0.0)
```
?

**Comment by TotalVerb** [comment]
I think it would be very confusing to mix real numbers and quantities with units. I would just define a single method for both arguments being `Dates.Period`.

**Comment by Shade5** [comment]
Done!

File: base/dates/types.jl [hunk 1]
```diff
@@ -241,3 +241,8 @@ Base.isless(x::Date,y::Date) = isless(value(x),value(y))
 Base.isless(x::DateTime,y::DateTime) = isless(value(x),value(y))
 Base.isless(x::TimeType,y::TimeType) = isless(promote(x,y)...)
 ==(x::TimeType,y::TimeType) = ===(promote(x,y)...)
+
+import Base: sleep,Timer,timedwait
+sleep(time::Dates.Period) = sleep(Dates.toms(time) / 1000)
```
Comment (tkelman) [inline base/dates/types.jl]: don't need `Dates.` while inside the `Dates` module

Comment (Shade5) [inline base/dates/types.jl]: My bad, Fixed it

File: base/dates/types.jl [hunk 2]
```diff
@@ -241,3 +241,8 @@ Base.isless(x::Date,y::Date) = isless(value(x),value(y))
 Base.isless(x::DateTime,y::DateTime) = isless(value(x),value(y))
 Base.isless(x::TimeType,y::TimeType) = isless(promote(x,y)...)
 ==(x::TimeType,y::TimeType) = ===(promote(x,y)...)
+
+import Base: sleep,Timer,timedwait
+sleep(time::Period) = sleep(toms(time) / 1000)
+Timer(time::Period, repeat::Period=0.0) = Timer(toms(time) / 1000,toms(repeat) / 1000)
```
Comment (TotalVerb) [inline base/dates/types.jl]: I don't think `0.0` will convert properly to `Period`. Use `Second(0)` as the default instead.

File: base/dates/types.jl [hunk 3]
```diff
@@ -241,3 +241,8 @@ Base.isless(x::Date,y::Date) = isless(value(x),value(y))
...
+timedwait(testcb::Function, time::Period, pollint::Float64) = timedwait(testcb, toms(time) / 1000, pollint)
```
Comment (TotalVerb) [inline base/dates/types.jl]: `pollint` is supposed to be a keyword argument; it's also a time.

**Review by TotalVerb — CHANGES_REQUESTED** [review]

File: base/dates/types.jl (hunk 3 above)
Comment (Shade5) [inline base/dates/types.jl]: Changed!

File: base/dates/types.jl [hunk 4]
```diff
@@ -241,3 +241,8 @@ Base.isless(x::Date,y::Date) = isless(value(x),value(y))
 Base.isless(x::DateTime,y::DateTime) = isless(value(x),value(y))
 Base.isless(x::TimeType,y::TimeType) = isless(promote(x,y)...)
 ==(x::TimeType,y::TimeType) = ===(promote(x,y)...)
+
+import Base: sleep,Timer,timedwait
+sleep(time::Period) = sleep(toms(time) / 1000)
+Timer(time::Period, repeat::Period=Second(0)) = Timer(toms(time) / 1000,toms(repeat) / 1000)
+timedwait(testcb::Function, time::Period, pollint::Period) = timedwait(testcb, toms(time) / 1000, toms(pollint) / 1000)
```
Comment (TotalVerb) [inline base/dates/types.jl]: The signature is still not correct. Keyword arguments should be delimited like

`timedwait(testcb, time::Period; pollint::Period=Millisecond(100)) = ...`

**Review by TotalVerb — CHANGES_REQUESTED** [review]

**Comment by TotalVerb** [comment]
Please add some tests that these methods work properly. See `test/channels.jl` for a good way to test timedwait; alternatively, it should also work just to test that these methods work at all.

**Comment by Shade5** [comment]
@TotalVerb Where are the default tests for sleep and Timer? I've tried, but I can't find them.

**Comment by TotalVerb** [comment]
There are no tests specifically for `sleep` and `Timer`, but it should be enough to test the `sleep(Dates.Millisecond(60))` version in `threads.jl`, replacing the existing `sleep(0.06)` calls.

**Comment by Shade5** [comment]
Alright, I'll add the tests, there seems to be a problem with timedwait. I've posted it on gitter.

**Comment by Shade5** [comment]
Added tests

**Comment by TotalVerb** [comment]
It would be better to avoid duplicating test code. Consider instead a `for` loop;

```julia
for period in (0.06, Dates.Millisecond(60))
    ... # testing code goes here, using `Timer(period)` / `sleep(period)`
end
```

**Comment by Shade5** [comment]
I didn't add the test for timedwait in a forloop as the log output was different

**Comment by Shade5** [comment]
Anything else?

File: test/channels.jl [hunk 5]
```diff
@@ -92,14 +92,14 @@ end
...
-    timedwait(callback, 1.0)
+    timedwait(callback, Dates.Second(1))
     et=toq()
     # assuming that 0.5 seconds is a good enough buffer on a typical modern CPU
     try
         @assert (et >= 1.0) && (et <= 1.5)
         @assert !isready(rr3)
     catch
-        warn("timedwait tests delayed. et=$et, isready(rr3)=$(isready(rr3))")
+        warn("timedwait-Period tests delayed. et=$et, isready(rr3)=$(isready(rr3))")
```
Comment (TotalVerb) [inline test/channels.jl]: No need for this `-Period` here.

**Review by TotalVerb — CHANGES_REQUESTED** [review]

**Comment by TotalVerb** [comment]
@tkelman I suppose since it is now past feature freeze, this will have to wait until 1.0, right?

File: test/channels.jl (hunk 5 above)
Comment (Shade5) [inline test/channels.jl]: :+1:

File: test/channels.jl [hunk 6]
```diff
@@ -92,7 +92,7 @@ end
     @async begin sleep(2.0); put!(rr3, :ok) end
 
     tic()
-    timedwait(callback, 1.0)
+    timedwait(callback, Dates.Second(1))
```
Comment (musm) [inline test/channels.jl]: why is this necessary, shouldn't `timedwait(callback, 1.0)` still work?

Comment (Shade5) [inline test/channels.jl]: As TotalVerb said, The Period version dispatches to the Real version anyway.

**Comment by TotalVerb** [comment]
Travis failure is https://github.com/JuliaLang/julia/issues/19792.

**Comment by Shade5** [comment]
@TotalVerb Should I do anything?

**Review by StefanKarpinski — APPROVED** [review]

**Pull request merged**